ocaml/database/db_cache_types.ml

```diff
@@ -126,6 +147,17 @@ functor
 module Row = struct
   include Make (Schema.Value)
 
+  let add gen key v =
+    add gen key
+    @@
+    match v with
+    | Schema.Value.String x ->
+        Schema.Value.String (unique x)
+    | Schema.Value.Pairs ps ->
+        Schema.Value.Pairs (List.map (fun (x, y) -> (unique x, unique y)) ps)
+    | Schema.Value.Set xs ->
+        Schema.Value.Set (List.map unique xs)
```

Comment:
Same here with `add_to_set`.
In fact the type of Pairs and Set is rather inefficient, the former should be a map, and the latter a true set. That would speed up some of the add/remove/lookup functions since they wouldn't have to walk a list anymore.
And potentially generate less garbage as well since removing a value wouldn't involve rewriting the whole list, but just updating up to a `log` depth on a tree.
Perhaps that could come as an optimization in a separate PR.

Reply:
I would like to keep this PR focused on the sharing aspect but I agree that the abundance of lists in a database at least looks inefficient.

Reply:
I have simplified the implementation. I am reluctant to modify `add_to_map` because it spreads out where we would add the sharing.